File: runtime/src/main/java/com/google/cloud/verticals/foundations/dataharmonization/imports/Parser.java
/*
 * Copyright 2020 Google LLC.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.cloud.verticals.foundations.dataharmonization.imports;

import com.google.cloud.verticals.foundations.dataharmonization.function.context.MetaData;
import com.google.cloud.verticals.foundations.dataharmonization.function.context.Registries;
import com.google.cloud.verticals.foundations.dataharmonization.function.context.RuntimeContext;
import com.google.cloud.verticals.foundations.dataharmonization.proto.Pipeline.PipelineConfig.Import;
import com.google.cloud.verticals.foundations.dataharmonization.registry.Registrable;
import java.io.IOException;

/**
 * A Parser parses the data loaded by a {@link Loader} and initializes the given {@link
 * RuntimeContext} with it. Implementations of this interface are referred to by {@link
 * Import#getParser()}
 */
public interface Parser extends Registrable {
  /**
   * Parse the given data and register it as appropriate into the given registries. If the given
   * data represents something that can import other things (like a Whistle file that can import
   * other Whistle files or plugins), then it should do so using the given ImportProcessor.
   */
  void parse(
      byte[] data,
      Registries registries,
      MetaData metaData,
      ImportProcessor processor,
      ImportPath iPath)
      throws IOException;

  /** Return true iff this parser can parse the file at the specified import path. */
  boolean canParse(ImportPath path);
}

File: wstl1/mapping_language/errors/parser.go
// Copyright 2019 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package errors

import (
	"fmt"

	"github.com/antlr/antlr4/runtime/Go/antlr" /* copybara-comment: antlr */
)

// ParserListener listens for and formats SyntaxErrors in a way suitable for
// the antlr Parser.
type ParserListener struct {
	*antlr.DefaultErrorListener
	Code string
}

// SyntaxError is called when the parser encounters a syntax error.
func (p *ParserListener) SyntaxError(recognizer antlr.Recognizer, offendingSymbol interface{}, line, column int, msg string, e antlr.RecognitionException) {
	if e != nil && e.GetOffendingToken() != nil {
		panic(NewTranspilationError(line, column, fmt.Errorf("parser error at %q: %s", e.GetOffendingToken().GetText(), msg)))
	}

	panic(NewTranspilationError(line, column, fmt.Errorf("parser error at %q: %s", nextToken(p.Code, line, column), msg)))
}
